apps/prairielearn/src/lib/client/downloads.ts

```diff
@@ -0,0 +1,61 @@
+/**
+ * Triggers a browser download of a text-based file.
+ *
+ * @param content The content of the file.
+ * @param filename The desired filename.
+ * @param mimeType The MIME type of the file.
+ */
+export function downloadTextFile(content: string, filename: string, mimeType: string): void {
+  if (typeof window === 'undefined') return;
+  const blob = new Blob([content], { type: mimeType });
+  const url = URL.createObjectURL(blob);
+  const a = document.createElement('a');
+  a.href = url;
+  a.download = filename;
+  document.body.appendChild(a);
+  a.click();
+  document.body.removeChild(a);
+  URL.revokeObjectURL(url);
+}
+
+/**
+ * Converts a 2D array into a CSV string.
+ *
+ * @param data 2D array of data.
+ * @returns A CSV-formatted string.
+ */
+export function arrayToCsv(data: unknown[][]): string {
+  return data
+    .map((row) =>
+      row
+        .map((field) => {
+          const stringField = String(field ?? '');
+          return /[",\n]/.test(stringField) ? `"${stringField.replace(/"/g, '""')}"` : stringField;
+        })
+        .join(','),
+    )
+    .join('\r\n');
```

Comment:
Manual CSV construction makes me a little nervous, but I could just be overestimating how difficult it is to correctly and safely produce a CSV string.

Should we reach for [`csv-stringify/sync`](https://github.com/adaltas/node-csv/blob/master/packages/csv-stringify) instead to let someone else deal with the complexity? At the very least, I'd encourage you to check out their implementation to see if we're missing any edge cases. If you want to continue rolling your own, we should have tests for this.

Reply:
Done in [c22ef32](https://github.com/PrairieLearn/PrairieLearn/pull/12278/commits/c22ef322227ecf117893a62f616be40bb4aa7c6f).